Sources/Valet/SecureEnclaveError.swift

```diff
@@ -0,0 +1,82 @@
+//  Created by Allison Moyer on 5/28/21.
+//  Copyright © 2021 Square, Inc. All rights reserved.
+//
+//  Licensed under the Apache License Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing software
+//  distributed under the License is distributed on an "AS IS" BASIS
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import LocalAuthentication
+
+public enum SecureEnclaveError: Int, CaseIterable, CustomStringConvertible, Error, Equatable {
+    /// Access to the secure enclave was not attempted because it was not available (due to biometrics not being available or enrolled
+    /// or being locked out, or the authentication context being invalidated)
+    case couldNotAccess
+
+    /// Access to the secure enclave was not attempted because the user failed to provide valid credentials.
+    case authenticationFailed
+
+    /// Access to the secure enclave was not attempted because the user canceled the prompt.
+    case userCancelled
+
+    /// Access to the secure enclave was not attempted because the user opted to the fallback option (i.e. 'Enter Password').
+    /// Callers should handle this as a custom alternative option to satisfying authentication.
+    case userFallback
+
+    /// Access to the secure enclave was not attempted because authentication was canceled by system
+    /// (e.g. another application went to foreground).
+    case systemCancel
+
+    /// Access to the secure enclave was not attempted because authentication could not start because
+    /// passcode is not set on the device
+    case passcodeNotSet
```

Comment:
I think this case is the same as `KeychainError.itemNotFound`, since if there's no passcode set there's no way that a biometric-protected item is stored in the secure enclave. Can we simplify?

Reply:
Sorry I think I wasn't clear! You're right about what this error message means. My point is more that if we were talking to the Keychain directly (rather than an `LAContext`), we'd have surfaced a `KeychainError.itemNotFound` error rather than something more tailored. I think there's value in 1. keeping the error codes aligned where possible and 2. simplifying the API.

I digress a bit, but I could imagine a future PR where we add an `var isPasscodeSet: Bool` method to a `SecureEnclaveValet` to better expose this particular configuration issue.

Reply:
Ohhh I see what you mean. Wouldn't that be the same argument for some of the `couldNotAccess` errors? If biometry isn't set up on the device, there won't be any items secured by it in the Keychain

Reply:
And would you suggest simplifying by returning `KeychainError.itemNotFound` or creating a `SecureEnclaveValet.itemNotFound`?

Reply:
> Wouldn't that be the same argument for some of the `couldNotAccess` errors?

I think so! If you think we can consolidate further I'd be happy to do so.

> If biometry isn't set up on the device, there won't be any items secured by it in the Keychain

Yes, and I think that's a pretty strong argument for utilizing `KeychainError.itemNotFound`, since we'll literally never be able to find a `SecureEnclaveValet` item in the keychain under this configuration.

The `couldNotAccessKeychain` error is more of a "the keychain is not accessible with your configuration" error... which I guess technically applies to this case, but historically it's been used more for "you gave us X accessibility but that condition hasn't been met yet" errors. It's nuanced, and I can see either error being appropriate here. But I lean towards `itemNotFound`, since unlike most `couldNotAccessKeychain` errors we _know_ there is no value in the keychain for this key.

> would you suggest simplifying by returning `KeychainError.itemNotFound` or creating a `SecureEnclaveValet.itemNotFound`?

The former I think is appropriate. I don't think we need a new `SecureEnclaveValet`-specific error class here.

Reply:
Sounds good!

Reply:
> But I lean towards `itemNotFound`, since unlike most `couldNotAccessKeychain` errors we _know_ there is no value in the keychain for this key.

The one potential downside is that `.itemNotFound` might imply that you _could_ set a value, whereas `.couldNotAccessKeychain` implies that there's no value due to the state of the keychain. Don't feel strongly either way though.

Reply:
That's a good point. Let's mirror what we'd do if a device were in this state and we talked directly to the Keychain rather than `LAContext` first. Which... I think that's `.itemNotFound` but I'm not 100% sure.

Given how we wrote `canAccessKeychain()` though, it would be `true` in this state. We may want to re-think this error condition and the APIs we surface to detect it 😅

Reply:
But we can do that re-thinking in another Issue/PR. Doesn't need to be this one 😅

Reply:
Oh yeah, you're totally right, that error should probably mirror `canAccessKeychain()`.

Reply:
Nope just oversight, thanks for catching!